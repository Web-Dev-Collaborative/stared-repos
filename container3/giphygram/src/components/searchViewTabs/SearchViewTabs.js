import React from 'react';
import PropTypes from 'prop-types';
import { SUPPORTED_COLUMNS_NUMS } from '../../services/LayoutService';
import './searchViewTabs.css';

const columnButtons = [
  { columns: 1, title: 'Show in 1 column', icon: 'oi oi-list' },
  { columns: 3, title: 'Show in 3 columns', icon: 'oi oi-grid-three-up' },
  { columns: 4, title: 'Show in 4 columns', icon: 'oi oi-grid-four-up' },
];

export class SearchViewTabs extends React.Component {
  static propTypes = {
    columnsNum: PropTypes.oneOf(SUPPORTED_COLUMNS_NUMS).isRequired,
    onColumnsNumChange: PropTypes.func.isRequired,
  };

  onButtonClick = (columnsNum) => {
    const { onColumnsNumChange } = this.props;
    onColumnsNumChange(columnsNum);
  };

  render() {
    const { columnsNum } = this.props;

    const buttons = columnButtons.map((columnButton) => {
      const buttonClass = columnButton.columns === columnsNum
        ? 'btn btn-dark tab-selector-button'
        : 'btn btn-light tab-selector-button';

      return (
        <button
          key={`button_${columnButton.columns}`}
          type="button"
          className={buttonClass}
          onClick={() => this.onButtonClick(columnButton.columns)}
          title={columnButton.title}
        >
          <span className={columnButton.icon} />
        </button>
      );
    });

    return (
      <div className="btn-group btn-group-sm mb-3">
        {buttons}
      </div>
    );
  }
}
