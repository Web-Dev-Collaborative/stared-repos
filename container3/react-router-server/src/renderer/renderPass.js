import React from 'react';
import { renderToString, renderToStaticMarkup } from 'react-dom/server';
import AsyncRenderer from '../components/AsyncRenderer';
import removeDuplicateModules from '../utils/removeDuplicateModules';

const renderPass = (context, element, staticMarkup = false) => {
  context.callback = () => {
    if (context.finishedLoadingModules && !context.statesRenderPass) {
      context.statesRenderPass = true;
      context.renderResult = renderPass(context, element, staticMarkup);
      if (context.fetchingStates <= 0 && context.modulesLoading <= 0) {
        context.resolve({
          html: context.renderResult,
          state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
          modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
        });
      }
    } else if (context.finishedLoadingModules && context.statesRenderPass || !context.hasModules) {
      context.renderResult = renderPass(context, element, staticMarkup);
      if (context.fetchingStates <= 0 && context.modulesLoading <= 0) {
        context.resolve({
          html: context.renderResult,
          state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
          modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
        });
      }
    }
  };

  let component = (
    <AsyncRenderer context={context}>
      {element}
    </AsyncRenderer>
  );

  let result;
  try {
    if (staticMarkup) {
      result = renderToStaticMarkup(component);
    } else {
      result = renderToString(component);
    }
  } catch (e) {
    return context.reject(e);
  }

  if (!context.hasModules && !context.hasStates) {
    context.resolve({
      html: result,
      state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
      modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
    });
  }
  return result;
};

export default renderPass;
