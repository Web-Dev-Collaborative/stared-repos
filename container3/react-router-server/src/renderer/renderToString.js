import renderPass from './renderPass';
import { clear } from '../module/cache';

export default (element) => {
  return new Promise((resolve, reject) => {
    clear();
    const context = {
      resolve, reject,
      modulesLoading: 0,
      fetchingStates: 0
    };
    renderPass(context, element);
  });
};
