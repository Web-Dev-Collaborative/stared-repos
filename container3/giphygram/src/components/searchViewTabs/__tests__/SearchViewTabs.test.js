import React from 'react';
import renderer from 'react-test-renderer';
import { SearchViewTabs } from '../SearchViewTabs';

describe('SearchViewTabs', () => {
  it('should be rendered correctly by default', () => {
    const onColumnsNumChange = jest.fn();

    const tree = renderer
      .create((
        <SearchViewTabs
          columnsNum={3}
          onColumnsNumChange={onColumnsNumChange}
        />
      ))
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should fire onColumnsNumbChange callback', () => {
    const onColumnsNumChange = jest.fn();

    const component = renderer
      .create((
        <SearchViewTabs
          columnsNum={3}
          onColumnsNumChange={onColumnsNumChange}
        />
      )).root;

    const buttons = component.findAllByType('button');

    expect(onColumnsNumChange).not.toHaveBeenCalled();
    expect(buttons).toBeDefined();

    // Imitate button clicks.
    buttons[0].props.onClick();
    buttons[1].props.onClick();
    buttons[2].props.onClick();

    expect(onColumnsNumChange).toHaveBeenCalledTimes(3);
    expect(onColumnsNumChange.mock.calls[0][0]).toBe(1);
    expect(onColumnsNumChange.mock.calls[1][0]).toBe(3);
    expect(onColumnsNumChange.mock.calls[2][0]).toBe(4);
  });
});
