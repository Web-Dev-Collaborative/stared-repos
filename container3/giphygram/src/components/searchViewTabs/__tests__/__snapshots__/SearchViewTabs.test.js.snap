// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchViewTabs should be rendered correctly by default 1`] = `
<div
  className="btn-group btn-group-sm mb-3"
>
  <button
    className="btn btn-light tab-selector-button"
    onClick={[Function]}
    title="Show in 1 column"
    type="button"
  >
    <span
      className="oi oi-list"
    />
  </button>
  <button
    className="btn btn-dark tab-selector-button"
    onClick={[Function]}
    title="Show in 3 columns"
    type="button"
  >
    <span
      className="oi oi-grid-three-up"
    />
  </button>
  <button
    className="btn btn-light tab-selector-button"
    onClick={[Function]}
    title="Show in 4 columns"
    type="button"
  >
    <span
      className="oi oi-grid-four-up"
    />
  </button>
</div>
`;
